packages/contracts/src/connect-contracts.ts

```diff
@@ -0,0 +1,141 @@
+import { Signer, Contract, providers, ethers } from 'ethers'
+import { Provider } from '@ethersproject/abstract-provider'
+import { getL1ContractData, getL2ContractData } from './contract-data'
+
+export type Network = 'goerli' | 'kovan' | 'mainnet'
+interface L1Contracts {
+  addressManager: Contract
+  canonicalTransactionChain: Contract
+  executionManager: Contract
+  fraudVerifier: Contract
+  xDomainMessenger: Contract
+  ethGateway: Contract
+  multiMessageRelayer: Contract
+  safetyChecker: Contract
+  stateCommitmentChain: Contract
+  stateManagerFactory: Contract
+  stateTransitionerFactory: Contract
+  xDomainMessengerProxy: Contract
+  l1EthGatewayProxy: Contract
+  mockBondManger: Contract
+}
+
+interface L2Contracts {
+  eth: Contract
+  xDomainMessenger: Contract
+  messagePasser: Contract
+  messageSender: Contract
+  deployerWhiteList: Contract
+  ecdsaContractAccount: Contract
+  sequencerEntrypoint: Contract
+  erc1820Registry: Contract
+  addressManager: Contract
+}
+
+/**
+ * Validates user provided a singer or provider & throws error if not
+ * @param signerOrProvider
+ */
+const checkSignerType = (signerOrProvider: Signer | Provider) => {
+  if (!signerOrProvider) {
+    throw Error('signerOrProvider argument is undefined')
+  }
+  if (
+    !Provider.isProvider(signerOrProvider) &&
+    !Signer.isSigner(signerOrProvider)
+  ) {
+    throw Error('signerOrProvider argument is the wrong type')
+  }
+}
+
+/**
+ * Connects a signer/provider to layer 1 contracts on a given network
+ * @param signerOrProvider ethers signer or provider
+ * @param network string denoting network
+ * @returns l1 contracts connected to signer/provider
+ */
+export const connectL1Contracts = async (
+  signerOrProvider,
+  network?: Network
+): Promise<L1Contracts> => {
+  checkSignerType(signerOrProvider)
+
+  if (!network) {
+    console.warn(
+      'network argument not provided to connectL1Contracts. Defaulting to mainnet.'
+    )
+    network = 'mainnet'
+  }
+  if (network !== 'mainnet' && network !== 'kovan' && network !== 'goerli') {
+    throw Error('network argument is the wrong type')
+  }
+
+  const l1ContractData = getL1ContractData(network)
+
+  const toEthersContract = (data) =>
```

Comment:
Nit: this function is defined twice

Reply:
It's technically 2 definitions. The signerOrProvider comes from outside its scope. 

I don't have a strong preference but Liam asked for several changes to minimize code. If the signerOrProvider is required as an argument, it results in a lot of repetition and ends up kind of becoming a pointless helper.

Reply:
Does making `signerOrProvider` be a optional arg help? ie `signerOrProvider?`